Assignment 4/assignment-4-submission-folder/battle.cpp

```diff
@@ -0,0 +1,202 @@
+void battle_turn(int playerIndex, Party* party, Enemy* enemy) {
+    cout << "Player " << playerIndex << ", choose your action: " << endl;
+    cout << "1. Attack" << endl;
+    cout << "2. Use an Item" << endl;
+    int action;
+    switch (action) {
+        case 1: // Attack
+            enemy->health -= party->players[playerIndex].damage;
+            cout << "Player " << playerIndex << " attacks and deals " << party->players[playerIndex].damage << " damage." << endl;
+            break;
+        case 2: // Use item
+            if (!player_has_items(&party->players[playerIndex])) {
+                cout << "You don't have any items." << endl;
+                battle_turn(playerIndex, part, enemy);
+            } else {
+                use_item(&party->players[playerIndex]);
+            }
+            break;
+        default:
+            cout << "Invalid Input. Please try again." << endl;
+            battle_turn(playerIndex, party, enemy);
+            break;
+    }
+}
+
+bool enemyDefeated = false;
+// Mineshaft battle
+void mineshaft_battle(Party* party) {
+    srand(time(nullptr));
+    int randIndex = rand() % MINESHAFT_ENEMIES;
+    Enemy_Type enemyType;
+    switch (randIndex) {
+        case 0:
+            enemyType = Shrimp;
+            break;
+        case 1:
+            enemyType = Mosquito;
+            break;
+        case 2:
+            enemyType = Mimic;
+            break;
+    }
+    Enemy* enemy = get_enemy_of_type(enemyType);
+    cout << "\nYou encounter a " << enemy->name << "\n" << endl;
+
+    // Battle loop
+    while (true) {
+        // Players turn
+        for (int i = 0; i < party->num_players; i++) {
+            if (party->players[i].health <= 0) {
+                continue;
+            }
+            battle_turn(i, party, enemy);
+
+        // Check if enemy is defeated
+            if (enemy->health <= 0) {
+                cout << "\nEnemy Defeated!\n" << endl;
+                enemyDefeated = true;
+                break;
+            }
+        }
+
+        // Enemy's turn
+        int targetIndex;
+        do {
+            targetIndex = rand() % party->num_players;
+        } while (party->players[targetIndex].health <= 0);
+        // Damage
+        party->players[targetIndex].health -= enemy->damage;
+        cout << "The " << enemy->name << " attacks player " << targetIndex << " and deals " << enemy->damage << " damage." << endl;
+        // Check if player is alive
+        if (party->players[targetIndex].health <= 0) {
+            cout << "Player " << targetIndex << " has been defeated!" << endl;
+        }
+
+        //Check if all players are defeated
+        bool allPlayersDefeated = true;
+        for (int i = 0; i < party->num_players; i++) {
+            if (party->players[i].health > 0) {
+                allPlayersDefeated = false;
+                break;
+            }
+        }
+        if (allPlayersDefeated) {
+            cout << "All players have been defeated!" << endl;
```

Comment:
game over is when all the players die; make sure to set `party->state` to `Game_State::Lost` (you'll likely have to include `player.h` for this)